### PR TITLE
Allow disabling of TTS for specific users

### PR DESCRIPTION
**Pull request by Popkornium18**
These changes make it possible to disable Text-To-Speech for specific
users while still receiving the messages as normal text.

A use case for this would be muting text to speech for a music bot.

### CODE REVIEW

File: src/mumble/Database.cpp [hunk 1]
```diff
@@ -238,6 +241,26 @@ void Database::setLocalIgnored(const QString &hash, bool ignored) {
...
+	return query.next();
+}
+
+void Database::setLocalIgnoredTTS(const QString &hash, bool ignoredTTS) {
+	QSqlQuery query(db);
+
+	if (ignoredTTS)
+		query.prepare(QLatin1String("INSERT INTO `ignored_tts` (`hash`) VALUES (?)"));
+	else
+		query.prepare(QLatin1String("DELETE FROM `ignored_tts` WHERE `hash` = ?"));
```
Comment (Krzmbrzl) [inline src/mumble/Database.cpp:259]: Does this query fail, if `hash` is not in the table? (Failing as in "crash the application")

File: src/mumble/Database.cpp [hunk 2]
```diff
@@ -238,6 +241,26 @@ void Database::setLocalIgnored(const QString &hash, bool ignored) {
...
+	query.addBindValue(hash);
+	execQueryAndLogFailure(query);
+	return query.next();
+}
+
+void Database::setLocalIgnoredTTS(const QString &hash, bool ignoredTTS) {
+	QSqlQuery query(db);
+
+	if (ignoredTTS)
+		query.prepare(QLatin1String("INSERT INTO `ignored_tts` (`hash`) VALUES (?)"));
```
Comment (Krzmbrzl) [inline src/mumble/Database.cpp:257]: Doesn't this allow for multiple entries in the table for the same `hash`?

File: src/mumble/MainWindow.cpp [hunk 3]
```diff
@@ -1585,6 +1586,7 @@ void MainWindow::qmUser_aboutToShow() {
 		qaUserLocalMute->setEnabled(false);
 		qaUserLocalVolume->setEnabled(false);
 		qaUserLocalIgnore->setEnabled(false);
+		qaUserLocalIgnoreTTS->setEnabled(false);
```
Comment (Krzmbrzl) [inline src/mumble/MainWindow.cpp:1589]: Maybe we only have to show this, if TTS is enabled in the first place?

File: src/mumble/Database.cpp (hunk 1 above)
Comment (Popkornium18) [reply src/mumble/Database.cpp:259]: I'll look into that.

It's done the same way as local muting and such. So if this code is problematic, these need to change too.

File: src/mumble/MainWindow.cpp (hunk 3 above)
Comment (Popkornium18) [reply src/mumble/MainWindow.cpp:1589]: Good point, I'll fix that.

File: src/mumble/Database.cpp (hunk 1 above)
Comment (Popkornium18) [reply src/mumble/Database.cpp:259]: No, it does not crash. I set the value, manually deleted it from sqlite and the unset it. Nothing happened.

File: src/mumble/Database.cpp (hunk 2 above)
Comment (Popkornium18) [reply src/mumble/Database.cpp:257]: No, there is a UNIQUE constraint on the index.

File: src/mumble/MainWindow.cpp (hunk 3 above)
Comment (Popkornium18) [reply src/mumble/MainWindow.cpp:1590]: Done

**Comment by streaps** [comment]
a screenshot would be helpful

**Comment by Popkornium18** [comment]
A screenshot of the option in the GUI?

**Comment by Popkornium18** [comment]
image

File: src/mumble/ClientUser.cpp [hunk 4]
```diff
@@ -190,6 +193,13 @@ void ClientUser::setLocalIgnore(bool ignore) {
 	emit muteDeafStateChanged();
 }
 
+void ClientUser::setLocalIgnoreTTS(bool ignoreTTS) {
+	if (bLocalIgnoreTTS == ignoreTTS)
+		return;
+	bLocalIgnoreTTS = ignoreTTS;
+	emit muteDeafStateChanged();
```
Comment (Krzmbrzl) [inline src/mumble/ClientUser.cpp:200]: I think you're misusing this signal here. After all TTS does have nothing to do with muting or deafening a user...

What is it that you want to achieve by emitting this signal here?

Comment (Popkornium18) [reply src/mumble/ClientUser.cpp:200]: Hmm, actually this method doesn't need to emit any signal. :thinking: I'll remove it.

**Comment by Popkornium18** [comment]
> Please have a look at our commit guidelines though: https://github.com/mumble-voip/mumble/blob/master/COMMIT_GUIDELINES.md
> If possible, it'd be nice if you could adapt your commits accoridngly (squash everything but the translation update into a single `FEAT(tts)` commit.

Ah shiet I used `FEAT(client)`, but that shouldn't be a problem, right? Also there are 2 new LOC, since I somehow missed to read the values from the DB, causing the checkbox to be always unchecked. (Messages.cpp:507)

**Comment by Krzmbrzl** [comment]
I changed the commit message to be `FEAT(tts)`. It's more specific than `FEAT(client)` :)

**Comment by Krzmbrzl** [comment]
Thank you for your contribution :)

**Comment by Krzmbrzl** [comment]
Just FYI: We have forgotten to check for null at some point - see #4307 :)

**Comment by rhetr** [comment]
has this feature been released yet? i don't see it in release 1.3.4 even though this was merged over a year ago.

**Comment by rhetr** [comment]
how do i use this feature? i don't see it in release 1.3.4 or in the latest 1.4.0 development snapshot.

**Comment by Krzmbrzl** [comment]
That it is not in the 1.3.x series is expected (we don't backport new features) but it should be present in the 1.4.0 snapshots. 
In fact this was included since the very first snapshot (see changelog at https://www.mumble.info/blog/first-mumble-1.4.0-development-snapshot/). Are you sure you are in fact starting the snapshot and not the 1.3.4 version?

**Comment by rhetr** [comment]
Screen Shot 2021-09-16 at 8 25 47 PM it says i'm using the development snapshot. i've tried on both windows and macos.

**Comment by Krzmbrzl** [comment]
The button will only show if you actually have TTS enabled in the first place. Is that the case?

https://github.com/mumble-voip/mumble/blob/937f4e2fd8657f6647b2cf0a1f78a45b4bfe44cd/src/mumble/MainWindow.cpp#L1568-L1569

**Pull request merged**